magicgui/_magicgui.py

```diff
@@ -1,18 +1,70 @@
 from __future__ import annotations
 
-from typing import TYPE_CHECKING, Callable, Optional, Union, overload
+import inspect
+from functools import partial
+from types import FunctionType
+from typing import (
+    TYPE_CHECKING,
+    Any,
+    Callable,
+    Generic,
+    Optional,
+    TypeVar,
+    Union,
+    overload,
+)
 from warnings import warn
 
 from typing_extensions import Literal
 
+from magicgui.widgets import FunctionGui, MainFunctionGui
+
 if TYPE_CHECKING:
     from magicgui.application import AppRef
-    from magicgui.widgets import FunctionGui, MainFunctionGui
+
+_T = TypeVar("_T")
+_R = TypeVar("_R")
+
+
+def _magicgui(function=None, factory=False, main_window=False, **kwargs):
+    """Actual private magicui decorator.
+
+    if factory is `True` will return a MagicFactory instance, that can be called
+    to return a `FunctionGui` instance.  See docstring of ``magicgui`` for parameters
```

Comment:
```suggestion
    to return a `FunctionGui` instance.  See docstring of ``magicgui`` for parameters.
    Otherwise, this will return a FunctionGui instance directly.
```